File: Lecture_1_C/problem_set/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int get_digit_sum(int number);

int main(void)
{
    // 获取用户输入
    long number;
    do
    {
        number = get_long("Number: ");
    }
    while (number < 0);

    // checksum
    int sum = 0;   // 按算法求和
    int digit = 0; // 数字位数
    int d1 = 0;    // 最高位
    int d2 = 0;    // 次高位
    while (number > 0)
    {
        digit++;
        d2 = d1;
        d1 = number % 10;
        sum += get_digit_sum(d1);
        number /= 10;
        if (number <= 0)
        {
            break;
        }
        digit++;
        d2 = d1;
        d1 = number % 10;
        sum += get_digit_sum(2 * d1);
        number /= 10;
    }

    // 结果判断
    if (sum % 10 == 0)
    {
        if (digit == 15 && d1 == 3 && (d2 == 4 || d2 == 7))
        {
            printf("AMEX\n");
        }
        else if (digit == 16 && d1 == 5 && d2 >= 1 && d2 <= 5)
        {
            printf("MASTERCARD\n");
        }
        else if ((digit == 13 || digit == 16) && d1 == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int get_digit_sum(int number)
{
    int sum = 0;
    while (number > 0)
    {
        sum += number % 10;
        number /= 10;
    }
    return sum;
}

File: Lecture_6_Python/problem_set/sentimental-credit/credit.py
def calculate_check_sum(number):
    sum = 0
    while number:
        sum += int(number % 10)
        number /= 10
    return sum


number = input("Number: ")
check_sum = 0
for i in range(len(number) - 1, -1, -2):
    check_sum += calculate_check_sum(int(number[i]))
    i -= 1
    if i >= 0:
        check_sum += calculate_check_sum(2 * int(number[i]))

if check_sum % 10 == 0:
    sub_number = number[0:2]
    if len(number) == 15 and (sub_number == "34" or sub_number == "37"):
        print("AMEX")
    elif len(number) == 16 and (sub_number == "51" or sub_number == "52" or sub_number == "53" or sub_number == "54" or sub_number == "55"):
        print("MASTERCARD")
    elif (len(number) == 13 or len(number) == 16) and number[0] == '4':
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")
